x/stake/keeper_test.go

```diff
@@ -259,36 +266,178 @@ func TestGetValidators(t *testing.T) {
 	assert.Equal(t, candidates[3].Address, validators[1].Address, "%v", validators)
 }
 
-// TODO
 // test the mechanism which keeps track of a validator set change
 func TestGetAccUpdateValidators(t *testing.T) {
-	//TODO
+	ctx, _, keeper := createTestInput(t, nil, false, 0)
+
+	amts := []int64{9, 8, 7, 10, 6}
+	var candidatesIn [5]Candidate
+	for i, amt := range amts {
+		candidatesIn[i] = Candidate{
+			Address:     addrVals[i],
+			PubKey:      pks[i],
+			Assets:      sdk.NewRat(amt),
+			Liabilities: sdk.NewRat(amt),
+		}
+	}
+
+	validatorsEqual := func(t *testing.T, expected []Validator, actual []Validator) {
+		require.Equal(t, len(expected), len(actual))
+		for i := 0; i < len(expected); i++ {
+			assert.Equal(t, expected[i], actual[i])
+		}
+	}
+
+	genValidators := func(candidates []Candidate) []Validator {
+		validators := make([]Validator, len(candidates))
+		for i, c := range candidates {
+			validators[i] = c.validator()
+		}
+		return validators
+	}
+
+	candidates := []Candidate{candidatesIn[1], candidatesIn[4]}
+	validators := genValidators(candidates)
+
 	// test from nothing to something
-	// test from something to nothing
+	acc := keeper.getAccUpdateValidators(ctx)
+	assert.Equal(t, 0, len(acc))
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	_ = keeper.GetValidators(ctx) // to init recent validator set
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test identical
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
+	acc = keeper.getAccUpdateValidators(ctx)
+
+	// test from something to nothing
+	keeper.removeCandidate(ctx, candidates[0].Address)
+	keeper.removeCandidate(ctx, candidates[1].Address)
+	acc = keeper.getAccUpdateValidators(ctx)
+	assert.Equal(t, 2, len(acc))
+	assert.Equal(t, validators[0].Address, acc[0].Address)
+	assert.Equal(t, int64(0), acc[0].VotingPower.Evaluate())
+	assert.Equal(t, validators[1].Address, acc[1].Address)
+	assert.Equal(t, int64(0), acc[1].VotingPower.Evaluate())
+
 	// test single value change
+	candidates[0].Assets = sdk.NewRat(600)
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test multiple value change
+	candidates[0].Assets = sdk.NewRat(200)
+	candidates[1].Assets = sdk.NewRat(0)
+	validators = genValidators(candidates)
+	keeper.setCandidate(ctx, candidates[0])
+	keeper.setCandidate(ctx, candidates[1])
+	acc = keeper.getAccUpdateValidators(ctx)
+	validatorsEqual(t, validators, acc)
+
 	// test validator added at the beginning
-	// test validator added in the middle
+	candidates = append([]Candidate{candidatesIn[0]}, candidates...)
```

Comment:
As per my previous comment we should not be adding to a list, but retrieving from the state